<div style="text-align: center;">
  <img src="https://github.com/Hack-io-Data/Imagenes/blob/main/01-LogosHackio/logo_naranja@4x.png?raw=true" alt="esquema" />
</div>

# Laboratorio Pandas (EDA y Unión de Datos)

## Los datos

Para este laboratorio, trabajarás con los siguientes conjuntos de datos:

- **`netflix_originals.csv`**: Contiene información sobre producciones originales de Netflix. Las columnas que nos encontraremos en este conjunto de datos son: 

   - `Title`: Contiene el nombre del título de la producción original de Netflix.

   - `Genre`: Especifica el género o combinación de géneros de la producción.

   - `Premiere`: Indica la fecha de estreno de la producción en Netflix.

   - `Runtime`: Contiene la duración de la producción en minutos.

   - `IMDB Score`: Representa la calificación promedio que la producción ha recibido en la plataforma IMDB.

   - `Language`: Especifica el idioma principal o los idiomas en los que se presenta la producción.

- **`netflix_titles.csv`**: Contiene información sobre una variedad de títulos disponibles en la plataforma Netflix, incluyendo tanto películas como series de televisión. Las columnas que nos encontraremos en este conjunto de datos son:

   - `show_id`: Identificador único para cada título en el dataset.

   - `type`: Especifica el tipo de contenido, ya sea "Movie" (película) o "TV Show" (serie de televisión).

   - `title`: Contiene el nombre del título de la producción.

   - `director`: Muestra el nombre del director del título. Puede estar vacío en algunos casos.

   - `cast`: Lista de actores y actrices que participaron en el título. También puede estar vacío en algunos casos.

   - `country`: País o países donde se produjo el título. Esta columna también puede tener valores faltantes.

   - `date_added`: Fecha en la que el título fue añadido al catálogo de Netflix.

   - `release_year`: Año en que el título fue lanzado o producido.

   - `rating`: Clasificación del contenido en términos de audiencia (por ejemplo, PG-13, TV-MA).

   - `duration`: Duración del título. Para películas, se expresa en minutos (e.g., "90 min") y para series en temporadas (e.g., "2 Seasons").

   - `listed_in`: Categorías o géneros bajo los cuales el título está clasificado (por ejemplo, "Documentaries", "TV Dramas").

   - `description`: Breve sinopsis o descripción del contenido del título.

A lo largo de los laboratorios de Pandas, trabajarás con estos conjuntos de datos (o con los derivados del Laboratorio), algunas de las actividades que realizaremos son:

- Exploración de datos y análisis exploratorio (EDA) para familiarizarse con los conjuntos de datos.

- Realización de uniones entre conjuntos de datos, aplicando diferentes técnicas de *merging* y analizando la relevancia de cada método.

- Uso de funciones de selección y filtrado (`loc`, `iloc`) para extraer información clave.

- Desarrollo de análisis descriptivos y visualización de patrones para comprender mejor las decisiones estratégicas de la plataforma.

----

## Objetivo General de la Serie de Laboratorios



1. **Unión de `netflix_titles.csv` con `netflix_originals.csv`:**

   - **Análisis de contenido original:**  Identificarás los títulos que son producciones originales de Netflix y los compararás con los títulos adquiridos. Este análisis te permitirá:

     - Comparar la diversidad de géneros entre los títulos originales y no originales.

     - Evaluar si ciertos géneros o tipos de contenido (como documentales o series) son más frecuentes en las producciones originales.

     - Investigar si la producción original se concentra en determinados países o idiomas.


   - **Evaluación de contenido:** En esta etapa, explorarás si las producciones originales de Netflix tienden a recibir mejores o peores evaluaciones en comparación con el contenido adquirido. Algunas preguntas que se buscarán responder incluyen:

     - ¿Los títulos originales tienen una calificación promedio superior en plataformas como IMDB?

     - ¿Son las producciones originales más consistentes en términos de calidad percibida?

     - ¿Existen diferencias en la popularidad de los títulos originales según su género o país de origen?


   - **Identificación de tendencias:**  Por último, este análisis te permitirá explorar cómo ha evolucionado la estrategia de contenido de Netflix a lo largo del tiempo. Algunos puntos clave a investigar serán:

     - ¿Cómo ha crecido la producción original en comparación con la adquisición de contenido?

     - ¿Qué géneros o tipos de contenido han recibido mayor inversión en los últimos años?
     
     - ¿Existen patrones en los países o idiomas de las producciones originales, especialmente en relación con la expansión global de Netflix?


----


## Ejercicio: Análisis y Enriquecimiento del Catálogo de Netflix

### Parte 1: Análisis Exploratorio de Datos (EDA)

1. **Carga de los datos:**

   - Deberás cargar los dos conjuntos de datos que te hemos proporcionado para realizar este Laboratorio:

     - `netflix_originals.csv`

     - `netflix_titles.csv`

   Deberás revisar la estructura de cada conjunto de datos utilizando los métodos aprendidos durante la lección.

2. **Análisis de columnas y consistencia:**

   - Deberás hacer un análisis exploratorio de cada uno de los conjuntos de datos. Será obligatorio que incluyas las conclusiones generales obtenidas de cada uno de los análisis exploratorios en el Jupyter Notebook de tus soluciones.

   - Explorar qué columnas contiene cada conjunto de datos y cuáles son sus tipos de datos. Identificar si hay valores nulos o duplicados.

   - Algunas preguntas que nos podemos plantear en esta fase son (recuerda que esto es solo un ejemplo, y que podemos explorar todo lo que queramos/necesitemos para entender lo mejor posible los conjuntos de datos):

      - ¿Cuáles son los tipos de datos de cada columna? Nos puede ayudar a detectar si algún tipo de dato está incorrectamente asignado, como fechas almacenadas como cadenas de texto.

      - ¿Existen valores nulos? ¿En qué columnas están? Este paso es crucial para decidir si deben ser eliminados, rellenados o ignorados.

      - ¿Cuántos títulos únicos hay en cada conjunto de datos? Nos puede permitir verificar si existen títulos duplicados que puedan distorsionar el análisis.

      - ¿Qué géneros tenemos en el conjunto de datos? Esto nos puede ayudar a identificar los géneros más populares o incluso a identificar una posible limpieza de dicha columna para poder extraer conclusiones más claras en nuestro análisis.

### Parte 2: Unión de Datos

1. **Realización de la unión:**

   - Deberás realizar la unión que consideres más apropiada entre `netflix_titles.csv` y `netflix_originals.csv`. Además, deberás justificar el tipo de unión que has realizado.

   - Esto nos ayudará a contestar preguntas más adelente del tipo:

     - ¿Cuántos títulos en `netflix_titles.csv` son producciones originales de Netflix?

     - ¿Hay diferencias significativas en la puntuación de IMDB entre títulos originales y no originales?

### Parte 3: Uso de `loc` e `iloc`

Usando el DataFrame creado en la Parte 2: 

1. **Filtrado de datos con `loc`:**

   - Selecciona todas las filas donde el tipo de contenido sea “Movie”.

   - Muestra solo los títulos (title) y la duración (duration) de todos los contenidos que sean de tipo “TV Show”.

   - Selecciona todas las filas donde el país sea “United States”.

   - Muestra todas las películas (type = “Movie”) y selecciona las columnas title y director.

   - Selecciona los títulos (title) y géneros (listed_in) de todas las producciones lanzadas en 2018.

   - Selecciona las filas donde la columna director esté vacía y muestra solo los títulos (title).

2. **Acceso a datos específicos con `iloc`:**

   - Selecciona las primeras 5 filas del conjunto de datos y muestra solo las columnas title, director y country.

   - Muestra las últimas 5 filas del conjunto de datos y selecciona las columnas title y description.

   - Selecciona las primeras 10 filas del dataset y muestra las columnas desde la segunda hasta la quinta (incluidas). 

   - Muestra las últimas 7 filas del dataset y selecciona las columnas title, director, y country usando índices posicionales.

   - Selecciona las filas desde la 15 hasta la 25 (incluidas) y muestra las primeras 4 columnas.

   - Muestra las filas 20 a 30 y solo las columnas title, release_year, y rating usando índices posicionales.


**NOTA IMPORTANTE**: Antes de terminar el lab, guarda el conjunto de datos creado en la fase de unión para trabajar con el en las siguientes lecciones. 

In [46]:
# Antes de empezar importamos la librería de NumPy para poder trabajar con todos sus métodos. 
import numpy as np
# ==============================================================================
# Tratamiento de datos
# ==============================================================================
import pandas as pd
# Ignorar warnings
# ==============================================================================
import warnings
warnings.filterwarnings("ignore")
# Configuración
# -----------------------------------------------------------------------
pd.set_option('display.max_columns', None) # para poder visualizar todas las columnas de los DataFrames

### Parte 1: Análisis Exploratorio de Datos (EDA)

In [47]:
## Cargamos los conjuntos de datos
df_titles = pd.read_csv("datos/netflix_titles.csv", index_col=0)
df_originals = pd.read_csv("datos/netflix_originals.csv", index_col=0)

**Análisis preliminar**

In [48]:
## Obtenemos las 3 primeras filas del conjunto netflix_originals para tener una idea de las columnas y los datos 
df_originals.head(3)

Title                  Genre           Premiere  Runtime  \
0  Enter the Anime            Documentary     August 5, 2019       58   
1      Dark Forces               Thriller    August 21, 2020       81   
2          The App  Science fiction/Drama  December 26, 2019       79   

   IMDB Score          Language  
0         2.5  English/Japanese  
1         2.6           Spanish  
2         2.6           Italian

In [49]:
## Calculamos el tamaño del conjunto de datos
df_originals.shape

(513, 6)

In [50]:
## Obtenemos las 3 primeras filas del conjunto netflix_titles para tener una idea de las columnas y los datos 
df_titles.head(3)

show_id     type                 title         director  \
0      s1    Movie  Dick Johnson Is Dead  Kirsten Johnson   
1      s2  TV Show         Blood & Water              NaN   
2      s3  TV Show             Ganglands  Julien Leclercq   

                                                cast        country  \
0                                                NaN  United States   
1  Ama Qamata, Khosi Ngema, Gail Mabalane, Thaban...   South Africa   
2  Sami Bouajila, Tracy Gotoas, Samuel Jouy, Nabi...            NaN   

           date_added  release_year rating   duration  \
0  September 25, 2021          2020  PG-13     90 min   
1  September 24, 2021          2021  TV-MA  2 Seasons   
2  September 24, 2021          2021  TV-MA        NaN   

                                           listed_in  \
0                                      Documentaries   
1    International TV Shows, TV Dramas, TV Mysteries   
2  Crime TV Shows, International TV Shows, TV Act...   

                                         description  
0  As her father nears the end of his life, filmm...  
1  After crossing paths at a party, a Cape Town t...  
2  To protect his family from a powerful drug lor...

In [51]:
## Calculamos el tamaño de ambos conjuntos de datos
print(f"El conjunto 'originals' tiene {df_originals.shape[0]} filas y {df_originals.shape[1]} columnas.")
print(f"El conjunto 'titles' tiene {df_titles.shape[0]} filas y {df_titles.shape[1]} columnas.")

El conjunto 'originals' tiene 513 filas y 6 columnas.
El conjunto 'titles' tiene 8807 filas y 12 columnas.


In [52]:
df_titles.loc[df_titles["title"] == "Enter the Anime"]

show_id   type            title       director  \
3623   s3624  Movie  Enter the Anime  Alex Burunova   

                                                   cast               country  \
3623  Kozo Morishita, Yoko Takahashi, Shinji Aramaki...  United States, Japan   

          date_added  release_year rating duration  \
3623  August 5, 2019          2019  TV-MA   59 min   

                                listed_in  \
3623  Documentaries, International Movies   

                                            description  
3623  What is anime? Through deep-dives with notable...

El conjunto netflix_titles (8807 filas y 12 columnas) es mucho más grande que el conjunto netflix_originals (513 filas y 6 columnas), de lo que se puede deducir que hay muy pocos originales dentro del catálogo total.
Al buscar uno de los títulos del archivo netflix_originals dentro del archivo netflix_titles vemos que aparece, por lo que se entiende que los originales aparecen en el catálogo total y que la columna title puede ser un punto de unión.

**Tipos de datos**

In [53]:
## obtenemos el tipo de datos para cada conjunto
df_originals.dtypes

Title          object
Genre          object
Premiere       object
Runtime         int64
IMDB Score    float64
Language       object
dtype: object

In [54]:
df_titles.dtypes

show_id         object
type            object
title           object
director        object
cast            object
country         object
date_added      object
release_year     int64
rating          object
duration        object
listed_in       object
description     object
dtype: object


Como las columnas que contienen datos de fecha no tienen el tipo de fecha, sino que son objetos, tenemos que convertirlas a formato datatime. Esto se hará para la columna "Premiere" en el conjunto originals y para la columna date_added en el conjunto titles. Después comprobamos que el tipo de dato es correcto.

In [55]:
df_originals["Premiere"] = pd.to_datetime(df_originals["Premiere"], format="mixed") 
## se ha usado 'format = mixed' por sugerencia de python al dar un error de formato porque una fecha aparecía con un punto en lugar de una coma.
df_titles["date_added"] = pd.to_datetime(df_titles["date_added"], format="mixed") 

In [56]:
df_originals.dtypes

Title                 object
Genre                 object
Premiere      datetime64[ns]
Runtime                int64
IMDB Score           float64
Language              object
dtype: object

In [57]:
df_titles.dtypes

show_id                 object
type                    object
title                   object
director                object
cast                    object
country                 object
date_added      datetime64[ns]
release_year             int64
rating                  object
duration                object
listed_in               object
description             object
dtype: object

In [58]:
## Añadimos una columna nueva en el conjunto originals para conservar la información de que son producciones originales.
df_originals["Original"] = "yes"

In [59]:
df_originals.columns

Index(['Title', 'Genre', 'Premiere', 'Runtime', 'IMDB Score', 'Language',
       'Original'],
      dtype='object')

**Análisis de nulos y duplicados**

In [60]:
## definimos una función que nos genere un reporte
def reporte(df):
    df_reporte = pd.DataFrame()
    df_reporte["valores_nulos"] = df.isnull().sum()
    df_reporte["porcentaje_nulos"] = round(((df.isnull().sum())/df.shape[0])*100,2)
    df_reporte["tipo_dato"]=df.dtypes
    return df_reporte

In [61]:
df_reporte_titles = reporte(df_titles)
df_reporte_titles

valores_nulos  porcentaje_nulos       tipo_dato
show_id                   0              0.00          object
type                      0              0.00          object
title                     0              0.00          object
director               2634             29.91          object
cast                    825              9.37          object
country                 831              9.44          object
date_added               10              0.11  datetime64[ns]
release_year              0              0.00           int64
rating                    4              0.05          object
duration               4813             54.65          object
listed_in                 0              0.00          object
description               0              0.00          object

In [62]:
df_reporte_originals = reporte(df_originals)
df_reporte_originals

valores_nulos  porcentaje_nulos       tipo_dato
Title                   0               0.0          object
Genre                   0               0.0          object
Premiere                0               0.0  datetime64[ns]
Runtime                 0               0.0           int64
IMDB Score              0               0.0         float64
Language                0               0.0          object
Original                0               0.0          object

En el conjunto 'originals' no hay valores nulos, sin embargo en el conjunto 'titles', sí que hay y en varias columnas, en porcentajes tan altos como el 54.65% en la duración o el 29.91% en el director.

In [63]:
## Comprobamos si hay duplicados
print(f"En el conjunto 'originals' hay {df_originals.duplicated().sum()} duplicados")
print(f"En el conjunto 'titles' hay {df_titles.duplicated().sum()} duplicados")

En el conjunto 'originals' hay 0 duplicados
En el conjunto 'titles' hay 0 duplicados


No hay registros duplicados en ninguno de los conjuntos.

**Análisis descriptivo**

In [64]:
## Como la mayoría de los datos del conjunto 'originals' son object, empezamos por la descripción de este tipo de datos
df_originals.describe(include = "object").T

count unique                                       top freq
Title      513    513  David Attenborough: A Life on Our Planet    1
Genre      513    106                               Documentary  132
Language   513     37                                   English  352
Original   513      1                                       yes  513

Para la columna "Title" no hay información relevante, ya que sabemos que no hay títulos duplicados y, en consecuencia, las columnas de top y freq no ofrecen información nueva.
La columna "Original" es la añadida anteriormente, por lo que tampoco es relevante.
En cuanto a las conlumnas "Genre" y "Languaje", se va a ampliar un poco la información.

In [65]:
## Obtenemos los 10 primeros géneros con mayor frecuencia
df_originals["Genre"].value_counts().head(10)

Genre
Documentary        132
Drama               73
Comedy              42
Romantic comedy     35
Thriller            33
Comedy-drama        14
Crime drama         10
Horror               9
Biopic               8
Action               7
Name: count, dtype: int64

In [66]:
## Obtenemos los 10 primeros idiomas con mayor frecuencia
df_originals["Language"].value_counts().head(10)

Language
English            352
Hindi               28
Spanish             26
French              18
Italian             14
Portuguese           9
Indonesian           9
Japanese             5
English/Spanish      5
Korean               5
Name: count, dtype: int64

Hay 106 géneros únicos, de los cuales el más frecuente es el documental, que se repite 132 veces, seguido por el drama (73), la comedia (42), comedia romántica (35) y thriller (33).
De los 37 idiomas, el más repetido es el inglés, en 352 de los 513 títulos, seguido del hindi (28) y el español (26).

In [67]:
## Obtenemos la descripción de los datos numéricos del conjunto 'originals'.
df_originals.describe().T

count                           mean                  min  \
Premiere      513  2019-06-16 01:21:24.210526208  2014-12-13 00:00:00   
Runtime     513.0                      94.674464                  4.0   
IMDB Score  513.0                       6.210916                  2.5   

                            25%                  50%                  75%  \
Premiere    2018-06-08 00:00:00  2019-10-18 00:00:00  2020-09-17 00:00:00   
Runtime                    87.0                 97.0                108.0   
IMDB Score                  5.6                  6.3                  6.9   

                            max        std  
Premiere    2021-05-27 00:00:00        NaN  
Runtime                   209.0  26.648121  
IMDB Score                  9.0    0.96885

De media, las produciones originales duran 95 minutos, siendo la mediana de 97; aunque hay producciones muy cortas (4 min) o muy largas (209). Respecto a la puntuación en IMDB, la media y la mediana son muy similares también, 6,21 y 6,3 respectivamente, con un mínimo de 2,5 y un máximo de 9.

In [68]:
## Al igual que en el conjunto 'originals', la mayoría de datos son de tipo object, por lo que empezamos por la descripción de este tipo de datos
df_titles.describe(include = "object").T

count unique                                                top  \
show_id      8807   8807                                              s8807   
type         8807      2                                              Movie   
title        8807   8807                                             Zubaan   
director     6173   4528                                      Rajiv Chilaka   
cast         7982   7692                                 David Attenborough   
country      7976    748                                      United States   
rating       8803     17                                              TV-MA   
duration     3994    185                                          2 Seasons   
listed_in    8807    514                       Dramas, International Movies   
description  8807   8775  Paranormal activity at a lush, abandoned prope...   

             freq  
show_id         1  
type         6131  
title           1  
director       19  
cast           19  
country      2818  
rating       3207  
duration      425  
listed_in     362  
description     4

### Parte 2: Unión de Datos

In [69]:
## Usamos rename para cambiar el nombre a la columna "Title" del conjunto 'originlas' para que se llame igual que la del conjunto 'title'
df_originals = df_originals.rename(columns = {"Title": "title"})

In [70]:
## Comprobamos que se ha actualizado correctamente el nombre de la columna.
df_originals.columns

Index(['title', 'Genre', 'Premiere', 'Runtime', 'IMDB Score', 'Language',
       'Original'],
      dtype='object')

In [71]:
## Como queremos que se conserven todas las filas del conjunto 'titles' usaremos un left merge, considerando el izquierdo el conjunto 'titles'
df_catalogo = df_titles.merge(df_originals, how='left', on ="title")

In [72]:
## df_originals tenía 513 filas y 7 columnas tras añadir la columna "Original" y df_titles tenía 8807 filas y 12 columnas. 
## Comprobamos que df_catalogo tenga 8807 filas y 18 columnas ya que la columna "title" debería solaparse
df_catalogo.shape

(8807, 18)

In [73]:
df_catalogo.head(1)


show_id   type                 title         director cast        country  \
0      s1  Movie  Dick Johnson Is Dead  Kirsten Johnson  NaN  United States   

  date_added  release_year rating duration      listed_in  \
0 2021-09-25          2020  PG-13   90 min  Documentaries   

                                         description        Genre   Premiere  \
0  As her father nears the end of his life, filmm...  Documentary 2020-10-02   

   Runtime  IMDB Score Language Original  
0     90.0         7.5  English      yes

In [74]:
## Guardamos el archivo df_catalogo
df_catalogo.to_csv("datos/catalogo_completo.csv")

In [75]:
df_catalogo[["duration", "Runtime", "listed_in", "Genre"]]

duration  Runtime                                          listed_in  \
0        90 min     90.0                                      Documentaries   
1     2 Seasons      NaN    International TV Shows, TV Dramas, TV Mysteries   
2           NaN      NaN  Crime TV Shows, International TV Shows, TV Act...   
3           NaN      NaN                             Docuseries, Reality TV   
4     2 Seasons      NaN  International TV Shows, Romantic TV Shows, TV ...   
...         ...      ...                                                ...   
8802    158 min      NaN                     Cult Movies, Dramas, Thrillers   
8803  2 Seasons      NaN             Kids' TV, Korean TV Shows, TV Comedies   
8804        NaN      NaN                            Comedies, Horror Movies   
8805        NaN      NaN                 Children & Family Movies, Comedies   
8806    111 min      NaN     Dramas, International Movies, Music & Musicals   

            Genre  
0     Documentary  
1             NaN  
2             NaN  
3             NaN  
4             NaN  
...           ...  
8802          NaN  
8803          NaN  
8804          NaN  
8805          NaN  
8806          NaN  

[8807 rows x 4 columns]

### Parte 3: Uso de `loc` e `iloc`

1. **Filtrado de datos con `loc`:**

In [80]:
##  Selecciona todas las filas donde el tipo de contenido sea “Movie”.
df_catalogo.loc[df_catalogo["type"] == "Movie"]

show_id   type                             title  \
0         s1  Movie              Dick Johnson Is Dead   
6         s7  Movie  My Little Pony: A New Generation   
7         s8  Movie                           Sankofa   
9        s10  Movie                      The Starling   
12       s13  Movie                      Je Suis Karl   
...      ...    ...                               ...   
8801   s8802  Movie                           Zinzana   
8802   s8803  Movie                            Zodiac   
8804   s8805  Movie                        Zombieland   
8805   s8806  Movie                              Zoom   
8806   s8807  Movie                            Zubaan   

                           director  \
0                   Kirsten Johnson   
6     Robert Cullen, José Luis Ucha   
7                      Haile Gerima   
9                    Theodore Melfi   
12              Christian Schwochow   
...                             ...   
8801                Majid Al Ansari   
8802                  David Fincher   
8804                Ruben Fleischer   
8805                   Peter Hewitt   
8806                    Mozez Singh   

                                                   cast  \
0                                                   NaN   
6     Vanessa Hudgens, Kimiko Glenn, James Marsden, ...   
7     Kofi Ghanaba, Oyafunmike Ogunlano, Alexandra D...   
9     Melissa McCarthy, Chris O'Dowd, Kevin Kline, T...   
12    Luna Wedler, Jannis Niewöhner, Milan Peschel, ...   
...                                                 ...   
8801  Ali Suliman, Saleh Bakri, Yasa, Ali Al-Jabri, ...   
8802  Mark Ruffalo, Jake Gyllenhaal, Robert Downey J...   
8804  Jesse Eisenberg, Woody Harrelson, Emma Stone, ...   
8805  Tim Allen, Courteney Cox, Chevy Chase, Kate Ma...   
8806  Vicky Kaushal, Sarah-Jane Dias, Raaghav Chanan...   

                                                country date_added  \
0                                         United States 2021-09-25   
6                                                   NaN 2021-09-24   
7     United States, Ghana, Burkina Faso, United Kin... 2021-09-24   
9                                         United States 2021-09-24   
12                              Germany, Czech Republic 2021-09-23   
...                                                 ...        ...   
8801                       United Arab Emirates, Jordan 2016-03-09   
8802                                      United States 2019-11-20   
8804                                      United States 2019-11-01   
8805                                      United States 2020-01-11   
8806                                              India 2019-03-02   

      release_year rating duration  \
0             2020  PG-13   90 min   
6             2021     PG      NaN   
7             1993  TV-MA  125 min   
9             2021  PG-13  104 min   
12            2021  TV-MA  127 min   
...            ...    ...      ...   
8801          2015  TV-MA      NaN   
8802          2007      R  158 min   
8804          2009      R      NaN   
8805          2006     PG      NaN   
8806          2015  TV-14  111 min   

                                             listed_in  \
0                                        Documentaries   
6                             Children & Family Movies   
7     Dramas, Independent Movies, International Movies   
9                                     Comedies, Dramas   
12                        Dramas, International Movies   
...                                                ...   
8801           Dramas, International Movies, Thrillers   
8802                    Cult Movies, Dramas, Thrillers   
8804                           Comedies, Horror Movies   
8805                Children & Family Movies, Comedies   
8806    Dramas, International Movies, Music & Musicals   

                                            description        Genre  \
0     As her father nears the end of his life, filmm...  Documentary   
6  

In [84]:
## Muestra solo los títulos (title) y la duración (duration) de todos los contenidos que sean de tipo “TV Show”.
df_catalogo.loc[df_catalogo["type"] == "TV Show",["title", "duration"]]

title   duration
1             Blood & Water  2 Seasons
2                 Ganglands        NaN
3     Jailbirds New Orleans        NaN
4              Kota Factory  2 Seasons
5             Midnight Mass        NaN
...                     ...        ...
8795        Yu-Gi-Oh! Arc-V  2 Seasons
8796             Yunus Emre  2 Seasons
8797              Zak Storm        NaN
8800     Zindagi Gulzar Hai        NaN
8803            Zombie Dumb  2 Seasons

[2676 rows x 2 columns]

In [85]:
## Selecciona todas las filas donde el país sea “United States”.
df_catalogo.loc[df_catalogo["country"] == "United States"]

show_id     type                 title         director  \
0         s1    Movie  Dick Johnson Is Dead  Kirsten Johnson   
9        s10    Movie          The Starling   Theodore Melfi   
15       s16  TV Show     Dear White People              NaN   
27       s28    Movie             Grown Ups     Dennis Dugan   
28       s29    Movie            Dark Skies    Scott Stewart   
...      ...      ...                   ...              ...   
8791   s8792    Movie           Young Adult    Jason Reitman   
8793   s8794    Movie  Yours, Mine and Ours     Raja Gosnell   
8802   s8803    Movie                Zodiac    David Fincher   
8804   s8805    Movie            Zombieland  Ruben Fleischer   
8805   s8806    Movie                  Zoom     Peter Hewitt   

                                                   cast        country  \
0                                                   NaN  United States   
9     Melissa McCarthy, Chris O'Dowd, Kevin Kline, T...  United States   
15    Logan Browning, Brandon P. Bell, DeRon Horton,...  United States   
27    Adam Sandler, Kevin James, Chris Rock, David S...  United States   
28    Keri Russell, Josh Hamilton, J.K. Simmons, Dak...  United States   
...                                                 ...            ...   
8791  Charlize Theron, Patton Oswalt, Patrick Wilson...  United States   
8793  Dennis Quaid, Rene Russo, Sean Faris, Katija P...  United States   
8802  Mark Ruffalo, Jake Gyllenhaal, Robert Downey J...  United States   
8804  Jesse Eisenberg, Woody Harrelson, Emma Stone, ...  United States   
8805  Tim Allen, Courteney Cox, Chevy Chase, Kate Ma...  United States   

     date_added  release_year rating duration  \
0    2021-09-25          2020  PG-13   90 min   
9    2021-09-24          2021  PG-13  104 min   
15   2021-09-22          2021  TV-MA      NaN   
27   2021-09-20          2010  PG-13      NaN   
28   2021-09-19          2013  PG-13      NaN   
...         ...           ...    ...      ...   
8791 2019-11-20          2011      R      NaN   
8793 2019-11-20          2005     PG      NaN   
8802 2019-11-20          2007      R  158 min   
8804 2019-11-01          2009      R      NaN   
8805 2020-01-11          2006     PG      NaN   

                                 listed_in  \
0                            Documentaries   
9                         Comedies, Dramas   
15                  TV Comedies, TV Dramas   
27                                Comedies   
28         Horror Movies, Sci-Fi & Fantasy   
...                                    ...   
8791  Comedies, Dramas, Independent Movies   
8793    Children & Family Movies, Comedies   
8802        Cult Movies, Dramas, Thrillers   
8804               Comedies, Horror Movies   
8805    Children & Family Movies, Comedies   

                                            description        Genre  \
0     As her father nears the end of his life, filmm...  Documentary   
9     A woman adjusting to life after a loss contend...          NaN   
15    Students of color navigate the daily slights a...          NaN   
27    Mourning the loss of their beloved junior high...          NaN   
28    A family’s idyllic suburban life shatters when...          NaN   
...                                                 ...          ...   
8791  When a divorced writer gets a letter from an o...          NaN   
8793  When a father of eight and a mother of 10 prep...          NaN   
8802  A political cartoonist, a crime reporter and a...          NaN   
8804  Looking to survive in a world taken over by zo...          NaN   
8805  Dragged from civilian life, a former superhero...          NaN   

       Premiere  Runtime  IMDB Score Language Original  
0    2020-10-02     90.0         7.5  English      yes  
9           NaT      NaN         NaN      NaN      NaN  
15          NaT      NaN         NaN      NaN      NaN  
27          NaT      NaN         NaN      NaN      NaN  
28          NaT      NaN         NaN      NaN      NaN 

In [86]:
## Muestra todas las películas (type = “Movie”) y selecciona las columnas title y director.
df_catalogo.loc[df_catalogo["type"] == "Movie",["title", "director"]]

title                       director
0                 Dick Johnson Is Dead                Kirsten Johnson
6     My Little Pony: A New Generation  Robert Cullen, José Luis Ucha
7                              Sankofa                   Haile Gerima
9                         The Starling                 Theodore Melfi
12                        Je Suis Karl            Christian Schwochow
...                                ...                            ...
8801                           Zinzana                Majid Al Ansari
8802                            Zodiac                  David Fincher
8804                        Zombieland                Ruben Fleischer
8805                              Zoom                   Peter Hewitt
8806                            Zubaan                    Mozez Singh

[6131 rows x 2 columns]

In [87]:
## Selecciona los títulos (title) y géneros (listed_in) de todas las producciones lanzadas en 2018.
df_catalogo.loc[df_catalogo["release_year"] == 2018,["title", "listed_in"]]

title  \
21                               Resurrection: Ertugrul   
37                                          Angry Birds   
73                                         King of Boys   
94                                            Show Dogs   
96    If I Leave Here Tomorrow: A Film About Lynyrd ...   
...                                                 ...   
8732                                      White Chamber   
8743                                           Wildlife   
8774                                          يوم الدين   
8775                                    Yeh Meri Family   
8803                                        Zombie Dumb   

                                              listed_in  
21    International TV Shows, TV Action & Adventure,...  
37                                Kids' TV, TV Comedies  
73                         Dramas, International Movies  
94                   Children & Family Movies, Comedies  
96                      Documentaries, Music & Musicals  
...                                                 ...  
8732                        Sci-Fi & Fantasy, Thrillers  
8743                         Dramas, Independent Movies  
8774   Dramas, Independent Movies, International Movies  
8775                International TV Shows, TV Comedies  
8803             Kids' TV, Korean TV Shows, TV Comedies  

[1147 rows x 2 columns]

In [96]:
## Selecciona las filas donde la columna director esté vacía y muestra solo los títulos (title).
df_catalogo.loc[df_catalogo["director"].isnull(),["title"]]

title
1                           Blood & Water
3                   Jailbirds New Orleans
4                            Kota Factory
10    Vendetta: Truth, Lies and The Mafia
14        Crime Stories: India Detectives
...                                   ...
8795                      Yu-Gi-Oh! Arc-V
8796                           Yunus Emre
8797                            Zak Storm
8800                   Zindagi Gulzar Hai
8803                          Zombie Dumb

[2634 rows x 1 columns]

2. **Acceso a datos específicos con `iloc`:**

In [98]:
## Selecciona las primeras 5 filas del conjunto de datos y muestra solo las columnas title, director y country.
df_catalogo.iloc[0:5, [2,3,5]]

title         director        country
0   Dick Johnson Is Dead  Kirsten Johnson  United States
1          Blood & Water              NaN   South Africa
2              Ganglands  Julien Leclercq            NaN
3  Jailbirds New Orleans              NaN            NaN
4           Kota Factory              NaN          India

In [100]:
## Muestra las últimas 5 filas del conjunto de datos y selecciona las columnas title y description.
df_catalogo.iloc[-5:, [2,11]]

title                                        description
8802       Zodiac  A political cartoonist, a crime reporter and a...
8803  Zombie Dumb  While living alone in a spooky town, a young g...
8804   Zombieland  Looking to survive in a world taken over by zo...
8805         Zoom  Dragged from civilian life, a former superhero...
8806       Zubaan  A scrappy but poor boy worms his way into a ty...

In [104]:
## Selecciona las primeras 10 filas del dataset y muestra las columnas desde la segunda hasta la quinta (incluidas). 
df_catalogo.iloc[0:10, 2:6]

title                       director  \
0              Dick Johnson Is Dead                Kirsten Johnson   
1                     Blood & Water                            NaN   
2                         Ganglands                Julien Leclercq   
3             Jailbirds New Orleans                            NaN   
4                      Kota Factory                            NaN   
5                     Midnight Mass                  Mike Flanagan   
6  My Little Pony: A New Generation  Robert Cullen, José Luis Ucha   
7                           Sankofa                   Haile Gerima   
8     The Great British Baking Show                Andy Devonshire   
9                      The Starling                 Theodore Melfi   

                                                cast  \
0                                                NaN   
1  Ama Qamata, Khosi Ngema, Gail Mabalane, Thaban...   
2  Sami Bouajila, Tracy Gotoas, Samuel Jouy, Nabi...   
3                                                NaN   
4  Mayur More, Jitendra Kumar, Ranjan Raj, Alam K...   
5  Kate Siegel, Zach Gilford, Hamish Linklater, H...   
6  Vanessa Hudgens, Kimiko Glenn, James Marsden, ...   
7  Kofi Ghanaba, Oyafunmike Ogunlano, Alexandra D...   
8  Mel Giedroyc, Sue Perkins, Mary Berry, Paul Ho...   
9  Melissa McCarthy, Chris O'Dowd, Kevin Kline, T...   

                                             country  
0                                      United States  
1                                       South Africa  
2                                                NaN  
3                                                NaN  
4                                              India  
5                                                NaN  
6                                                NaN  
7  United States, Ghana, Burkina Faso, United Kin...  
8                                     United Kingdom  
9                                      United States

In [105]:
## Muestra las últimas 7 filas del dataset y selecciona las columnas title, director, y country usando índices posicionales.
df_catalogo.iloc[-7:, [2,3,5]]

title         director                       country
8800  Zindagi Gulzar Hai              NaN                      Pakistan
8801             Zinzana  Majid Al Ansari  United Arab Emirates, Jordan
8802              Zodiac    David Fincher                 United States
8803         Zombie Dumb              NaN                           NaN
8804          Zombieland  Ruben Fleischer                 United States
8805                Zoom     Peter Hewitt                 United States
8806              Zubaan      Mozez Singh                         India

In [106]:
## Selecciona las filas desde la 15 hasta la 25 (incluidas) y muestra las primeras 4 columnas.
df_catalogo.iloc[15:26, :4]

show_id     type                                              title  \
15     s16  TV Show                                  Dear White People   
16     s17    Movie  Europe's Most Dangerous Man: Otto Skorzeny in ...   
17     s18  TV Show                                    Falsa identidad   
18     s19    Movie                                          Intrusion   
19     s20  TV Show                                             Jaguar   
20     s21  TV Show    Monsters Inside: The 24 Faces of Billy Milligan   
21     s22  TV Show                             Resurrection: Ertugrul   
22     s23    Movie                                    Avvai Shanmughi   
23     s24    Movie       Go! Go! Cory Carson: Chrissy Takes the Wheel   
24     s25    Movie                                              Jeans   
25     s26  TV Show                               Love on the Spectrum   

                                         director  
15                                            NaN  
16  Pedro de Echave García, Pablo Azorín Williams  
17                                            NaN  
18                                     Adam Salky  
19                                            NaN  
20                                Olivier Megaton  
21                                            NaN  
22                                 K.S. Ravikumar  
23                        Alex Woo, Stanley Moore  
24                                     S. Shankar  
25                                            NaN

In [107]:
## Muestra las filas 20 a 30 y solo las columnas title, release_year, y rating usando índices posicionales.
df_catalogo.iloc[20:31, [2,7,8]]

title  release_year rating
20  Monsters Inside: The 24 Faces of Billy Milligan          2021  TV-14
21                           Resurrection: Ertugrul          2018  TV-14
22                                  Avvai Shanmughi          1996  TV-PG
23     Go! Go! Cory Carson: Chrissy Takes the Wheel          2021   TV-Y
24                                            Jeans          1998  TV-14
25                             Love on the Spectrum          2021  TV-14
26                                   Minsara Kanavu          1997  TV-PG
27                                        Grown Ups          2010  PG-13
28                                       Dark Skies          2013  PG-13
29                                         Paranoia          2013  PG-13
30                                  Ankahi Kahaniya          2021  TV-14